lib/ABI/FunctionType-revng-c.cpp

```diff
@@ -0,0 +1,424 @@
+  tryConvertingReturnValue(const SortedVector<RegisterType> &UsedRegisters);
+};
+
+class Converter : public ConverterBase {
+public:
+  using ConverterBase::ConverterBase;
+  TypeDependencyManager manageTypeDependencies(model::Binary &Binary) const {
+    return TypeDependencyManager(ConverterBase::manageTypeDependencies(Binary));
+  }
+
+public:
+  /// Entry point for the `toCABI` conversion.
+  std::optional<model::TypePath> toCABI(const model::RawFunctionType &Function,
+                                        TupleTree<model::Binary> &Binary) const;
+};
+
+std::optional<model::TypePath>
+Converter::toCABI(const model::RawFunctionType &Function,
+                  TupleTree<model::Binary> &Binary) const {
+  revng_log(Log, "Converting a `RawFunctionType` to `CABIFunctionType`.");
+  revng_log(Log, "Function:\n" << serializeToString(Function));
+
+  if (!ABI.checkCompatibility(Function.Arguments(), Function.ReturnValues())) {
+    revng_log(Log,
+              "FAIL: the function is not compatible with `"
+                << model::ABI::getName(ABI.ABI()) << "`.");
+    return std::nullopt;
+  }
+
+  // A single place to conveniently store all the types that need to be added
+  // to the model in order for the new function prototype to be valid.
+  auto M = manageTypeDependencies(*Binary);
+
+  // Convert register arguments first.
+  auto RegArguments = M.tryConvertingRegisterArguments(Function.Arguments());
+  if (RegArguments == std::nullopt) {
+    revng_log(Log, "FAIL: Register argument conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // Then stack arguments.
+  auto Stack = M.tryConvertingStackArguments(Function.StackArgumentsType(),
+                                             RegArguments->size());
+  if (Stack == std::nullopt) {
+    revng_log(Log, "FAIL: Stack argument conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // And the return type.
+  auto ReturnType = M.tryConvertingReturnValue(Function.ReturnValues());
+  if (ReturnType == std::nullopt) {
+    revng_log(Log, "FAIL: Return value conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // If all the partial conversions passed, the conversion is guaranteed to be
+  // successful. All the `TypeDependencies` can now be recorded into the
+  // model.
+  M.commit();
+
+  // With all the dependencies out of the way, new `CABIFunctionType` can now
+  // be created,
+  auto [NewType, NewTypePath] = Binary->makeType<model::CABIFunctionType>();
+  NewType.ID() = Function.ID();
+  NewType.CustomName() = Function.CustomName();
+  NewType.OriginalName() = Function.OriginalName();
+  NewType.ABI() = ABI.ABI();
+
+  // And filled in with the argument information.
+  using model::Argument;
+  for (auto &Argument : llvm::concat<Argument>(*RegArguments, *Stack))
+    NewType.Arguments().insert(std::move(Argument));
+  NewType.ReturnType() = *ReturnType;
+
+  revng_log(Log, "Conversion successful.");
+
```

Comment:
Can you add `TODO` (and add it to your list of things we'd like to have) for asserting that `FunctionType::Layout` of the old and new functions are identical (in case we the `ProduceCompatible` flag I suggested is disabled) or at least compatible.

```c++
revng_assert(FunctionType::Layout::make(NewType).isCompatibleWith(FunctionType::Layout::make(OldType)));
```

Do not do this now!

Reply:
I'm already doing something very similar as the part of the regular testing!

It's more fruitful to do it there because there's a lot of extra information available (we know for sure which argument is where and the like, thanks for `quemu`-dumping).